File: python/trees/subtree-of-another-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isSubtree(self, root: Optional[TreeNode], subRoot: Optional[TreeNode]) -> bool:
        candidates = []

        def checker(node):
            if not node:
                return

            if node.val == subRoot.val:
                candidates.append(node)

            checker(node.left)
            checker(node.right)

        checker(root)

        def dfs(p, q):
            if not p and not q:
                return True
            if not p or not q:
                return False

            if p.val != q.val:
                return False

            return dfs(p.left, q.left) and dfs(p.right, q.right)

        for i in candidates:
            if dfs(i, subRoot):
                return True

        return False

File: c++/trees/subtree-of-another-tree.cc
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isSubtree(TreeNode* root, TreeNode* subRoot) {
        std::queue<TreeNode*> q;
        q.push(root);
        while (q.size() != 0) {
            TreeNode* node = q.front();
            q.pop();

            if (!node) {
                continue;
            }

            if (node->val == subRoot->val) {
                if (compare(node, subRoot)) {
                    return true;
                }
            }

            q.push(node->left);
            q.push(node->right);
        }
        return false;
    }

private:
    bool compare(TreeNode* p, TreeNode* q) {
        if (!p && !q) {
            return true;
        }

        if (!p || !q) {
            return false;
        }

        return p->val == q->val && compare(p->left, q->left) && compare(p->right, q->right);
    }
};
